users/luescher/setups/hybrid/gmm_system.py

```diff
@@ -0,0 +1,845 @@
+            },
+        )
+        self.feature_flows[corpus][name] = features.basic_cache_flow(feature_path)
+        self.feature_flows[corpus][f"uncached_{name}"] = feature_job.feature_flow
+
+    @tk.block()
+    def extract_features(self, feat_args: dict, **kwargs):
+        corpus_list = self.train_corpora + self.dev_corpora + self.test_corpora
+
+        if "mfcc" in feat_args.keys():
+            for c in corpus_list:
+                self.mfcc_features(c, **feat_args["mfcc"])
+                self.energy_features(c, **feat_args["energy"])
+            for t in self.train_corpora:
+                self.add_energy_to_features(t, "mfcc+deriv")
+
+        if "gt" in feat_args.keys():
+            for c in corpus_list:
+                self.gt_features(c, **feat_args["gt"])
+                self.energy_features(c, **feat_args["energy"])
+            for t in self.train_corpora:
+                self.add_energy_to_features(t, "gt")
+
+        if "fb" in feat_args.keys():
+            for c in corpus_list:
+                self.fb_features(c, **feat_args["fb"])
+
+        unknown_features = set(feat_args.keys()).difference(
+            {"mfcc", "gt", "fb", "energy"}
+        )
+        if len(unknown_features) > 0:
+            raise ValueError("Invalid features: {}".format(unknown_features))
+
+    # -------------------- Mono Training --------------------
+
+    @tk.block()
+    def monophone_training(
+        self,
+        name,
+        corpus,
+        linear_alignment_args,
+        feature_energy_flow,
+        feature_flow,
+        align_iter,
+        splits,
+        accs_per_split,
+        align_keep_values=None,
+        **kwargs,
+    ):
+        self.linear_alignment(
+            name, corpus, feature_energy_flow, **linear_alignment_args
+        )
+
+        action_sequence = meta.align_and_accumulate_sequence(
+            align_iter, 1, mark_accumulate=False, mark_align=False
+        )
+        action_sequence += meta.split_and_accumulate_sequence(
+            splits, accs_per_split
+        ) + ["align!"]
+
+        akv = dict(**self.default_align_keep_values)
+        if align_keep_values is not None:
+            akv.update(align_keep_values)
+
+        self.train(
+            name=name,
+            corpus=corpus,
+            sequence=action_sequence,
+            flow=feature_flow,
+            initial_mixtures=meta.select_element(
+                self.mixtures, corpus, "linear_alignment_{}".format(name)
+            ),
+            align_keep_values=akv,
+            **kwargs,
+        )
+        self.jobs[corpus]["train_{}".format(name)].selected_alignment_jobs[
+            -1
+        ].add_alias("train_{}_alnt_last".format(name))
+        self.jobs[corpus]["train_{}".format(name)].selected_mixture_jobs[-1].add_alias(
+            "train_{}_mix_last".format(name)
+        )
+        tk.register_output(
+            "train_{}_{}_alnt_bundle_last".format(corpus, name),
```

Comment:
I don't like "alnt" as abbreviation for alignment and would prefer 'align' or full 'alignment'. But that's purely cosmetically.